flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/HsMemoryDataManager.java

```diff
@@ -0,0 +1,673 @@
+    private void releaseBuffers(Map<Integer, List<BufferIndexAndChannel>> toRelease) {
+        toRelease.forEach(
+                (subpartitionId, buffers) -> {
+                    HsSubpartitionDataBuffer subpartitionDataBuffer =
+                            getTargetSubpartitionDataBuffer(subpartitionId);
+                    final Object subpartitionLock = subpartitionDataBuffer.getSubpartitionLock();
+                    LockUtils.runWithLock(
+                            lock.readLock(),
+                            () -> {
+                                synchronized (subpartitionLock) {
+                                    buffers.forEach(
+                                            (indexAndChannel) -> {
+                                                int bufferIndex = indexAndChannel.getBufferIndex();
+                                                BufferContext bufferContext =
+                                                        subpartitionDataBuffer.getBufferContext(
+                                                                bufferIndex);
+                                                // only spill and not consumed buffer needs to be
+                                                // marked as readable.
+                                                if (subpartitionDataBuffer.isBufferSatisfyStatus(
+                                                        bufferContext,
+                                                        SpillStatus.SPILL,
+                                                        ConsumeStatus.NOT_CONSUMED)) {
+                                                    checkNotNull(
+                                                                    bufferContext.spilledFuture,
+                                                                    "Buffer in spill status should already set spill future.")
+                                                            .thenRun(
+                                                                    () ->
+                                                                            dataIndex
+                                                                                    .markBufferReadable(
+                                                                                            subpartitionId,
+                                                                                            bufferIndex));
+                                                }
+                                                subpartitionDataBuffer.onBufferReleased(
+                                                        bufferIndex);
+                                            });
+                                }
+                            });
+                });
+    }
+
+    private void recycleBuffer(MemorySegment buffer) {
+        numRequestedBuffers.decrementAndGet();
+        bufferPool.recycle(buffer);
+    }
+
+    class HsSubpartitionDataBuffer {
```

Comment:
I'd suggest not to make this an inner class, for a better separation between `HsMemoryDataManager` and `HsSubpartitionDataBuffer`.

Reply:
```suggestion
    class HsSubpartitionMemoryDataManager {
```

Reply:
done.